[MRG + 2] Refactor kneighbors_graph and radius_neighbors_graph (et.al)

MechCoder opened the pull request:
Following the discussion in https://github.com/scikit-learn/scikit-learn/pull/4019 this PR
1. Allows X to be None in `neighbors.kneighbors` which skips the first neighbor in both distance and connectivity mode.
2. Do not do anything special in other cases.

```
In [2]: knn = NearestNeighbors(n_neighbors=1).fit([[0], [1]])
In [3]: knn.kneighbors_graph([[0], [1]]).toarray()
Out[3]: 
array([[ 1.,  0.],
       [ 0.,  1.]])
In [5]: knn.kneighbors_graph([[2], [1]]).toarray()
Out[5]: 
array([[ 0.,  1.],
       [ 0.,  1.]])
In [8]: knn.kneighbors_graph().toarray()
Out[8]: 
array([[ 0.,  1.],
       [ 1.,  0.]])
In [6]: knn.kneighbors_graph([[2], [1]], mode='distance').toarray()
Out[6]: 
array([[ 0.,  1.],
       [ 0.,  0.]])
In [7]: knn.kneighbors_graph([[0], [1]], mode='distance').toarray()
Out[7]: 
array([[ 0.,  0.],
       [ 0.,  0.]])
In [9]: knn.kneighbors_graph(mode='distance').toarray()
Out[9]: 
array([[ 0.,  1.],
       [ 1.,  0.]])
```

Also, a happy 2015 to everyone!

MechCoder committed:
FIX/ENH: Fix kneighbors_graph/kneighbors and allow X=None

MechCoder commented:
ping @jnothman @jakevdp

MechCoder commented:
I have added a `DeprecationWarning` to account for the change in connectivity mode. Is that sufficient?

jnothman commented on sklearn/neighbors/graph.py:
```diff
@@ -77,7 +79,12 @@ def kneighbors_graph(X, n_neighbors, mode='connectivity', metric='minkowski',
+    if mode == "connectivity":
+        warnings.warn("The behavior of marking each sample itself as the "
+                      "first nearest neighbor will be changed in 0.18",
+                      DeprecationWarning)
```
Probably `FutureWarning`?

jnothman commented:
I don't think there are currently any tests for `NearestNeighbors().fit(X1).kneighbors_graph(X2)` and similar.

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -391,20 +400,45 @@ def kneighbors_graph(self, X, n_neighbors=None,
+            # is the sample itself. The same argument holds true for mode=distance
+            if not new_graph:
+                A_ind = self.kneighbors(X, n_neighbors + 1, return_distance=False)
+                # First column coressponds to the diagonal
```
This is not necessarily true when there are duplicate points.

jnothman replied:
And should be tested!

MechCoder replied:
I thought we need not bother about that.

jnothman replied:
Then you're going to get weird connectivity results...

MechCoder replied:
But what is wrong with the current behavior, it treats the duplicates as new samples.

    In [14]: from sklearn.neighbors import NearestNeighbors
    In [15]: X = np.random.rand(5, 5)
    In [16]: X[1], X[2] = X[0], X[0]
    In [17]: nn = NearestNeighbors(n_neighbors=2).fit(X).kneighbors_graph().A
    In [18]: nn
    Out[18]: 
    array([[ 0.,  1.,  1.,  0.,  0.],
           [ 0.,  1.,  1.,  0.,  0.],
           [ 0.,  1.,  1.,  0.,  0.],
           [ 0.,  1.,  1.,  0.,  0.],
           [ 0.,  0.,  1.,  1.,  0.]])

MechCoder replied:
Ah I see what you mean, setting the diagonal to zero is the way to go forward.

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -391,20 +400,45 @@ def kneighbors_graph(self, X, n_neighbors=None,
+                zero_data = A_data != 0
+
+                if not np.all(zero_data):
+                # Corner case, we need to rewrite data, ind and indptr
```
Why? You can just construct the matrix and call `eliminate_zeros` on it...

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -379,11 +379,20 @@ def kneighbors_graph(self, X, n_neighbors=None,
             n_neighbors = self.n_neighbors
 
+        if X is None:
+            new_graph = False
```
I think you may be best off setting `n_neighbors += 1` and then deleting the diagonal. In the connectivity case, just make `A_data` of ones before the deletion, and you can share most of the implementation regardless of `mode`.

MechCoder replied:
Any specific reason you want me to set `n_neighbors += 1` and then delete the diagonal? I have to do the `if-else` in that case too to set the diagonal elements to zero.

I'm also not sure about refactoring the implementation regardless of mode, without setting `return_distance=True` in both cases? I would need an if-else here also to unpack the tuple.

MechCoder replied:
This will also fail if there are duplicates. I cannot think of a proper way to handle duplicates.

MechCoder replied:
Wait, should the behavior of kneighbors also be modified? So that it does not mark the sample as the first nearest neighbor when X is None?

jnothman replied:
I didn't mean you could completely factor out the mode case, but you could avoid doing sub-conditions on `new_graph`  (which, I might add, is a confusing variable name).

jnothman commented:
Should this change in behaviour actually affect `kneighbors`, not just `_graph`? And what of `radius_neighbors_graph`?

MechCoder commented:
I asked the same question over here, https://github.com/scikit-learn/scikit-learn/pull/4046#discussion_r22437533

MechCoder commented:
I think we need to extend the same behavior to `kneighbors` and `radius_neighbors` as well.

MechCoder commented:
@jnothman @jakevdp I've addressed all your comments and have added a good number of tests. Would be great, if you could have a look.

jnothman commented:
I've not looked at your latest changes yet, but I would like to consider a tricky case. Say there are duplicates in the data, and I call the `X=None` variant of `radius_neighbors_graph`. I think I would like those duplicate points to be *explicit zeros* in the sparse matrix, so I can still extract the kth nearest distance from among the radius-neighbors (as required in OPTICS clustering). WDYT? Should this be tested?

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -300,11 +300,26 @@ class from an array representing our data set and ask who's
+            # itself as the first neighbor.
+            if n_neighbors >= X.shape[0]:
+                raise ValueError(
+                    "The number of neighbors should be at least one less"
```
This is confusing. How about "Expected n_neighbors <= {} for index of {} samples. Got {}." But why is a similar error not applicable to the `query_is_train=False` case?

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -315,11 +330,16 @@ class from an array representing our data set and ask who's
                                           **self.effective_metric_params_)
 
-            neigh_ind = argpartition(dist, n_neighbors - 1, axis=1)
+            neigh_ind = argpartition(dist, n_neighbors-1, axis=1)
```
why this change to spacing?

jnothman commented:
Two issues:
* there is a really bad case where there are duplicate points in the dataset: `X = np.ones((3, 1)); NearestNeighbors(n_neighbors=1).fit(X).kneighbors(X, return_distance=False)` will return an array of zeros. You're not going to find all the indices of the query points in there if X is None, so you'll get an error in the reshape! The solution is to mask the first entry where your mask matches nothing (but because the method is now more complex, I suggest you refactor a bit between brute and tree implementations).
* now I guess anything inheriting from `KNeighborsMixin` should also support the `X=None` case, which means `LSHForest`. This is bad because DRY.

amueller commented:
I feel the proposed interface is a bit counterintuitive. What is the reason to choose ``X=None`` for this?

jnothman commented:
Which old option? We're fixing a bug, and an issue that applies only if the training data is the test data (but not merely if it happens to contain queries from the same set). This special semantics needs some kind of special interface. It doesn't need to be `X=None`, but there are too many methods to duplicate them all, so it would need to be another parameter, but then what does X mean? And knowing that the query is the index data is useful for other things, like more efficient query (not implemented yet).

amueller commented:
I meant the "skip first" option. The bug that you mean is the inconsistency in #4015, right?
I should probably reread the fix again.
Knowing that the data is the training data is indeed useful information.

jnothman commented:
The skip first option is not really skip first. The idea is that if I have
a dataset X and want to find the nearest neighbors in X for each x in X,
then asking for 1NN shouldn't return me the query item. However, if I want
to find the nearest neighbors in X for each x in X2, then asking for the
1NN should return whatever is nearest in X.

On 8 January 2015 at 09:53, Andreas Mueller <notifications@github.com>
wrote:

>  I meant the "skip first" option. The bug that you mean is the
> inconsistency in #4015
> <https://github.com/scikit-learn/scikit-learn/issues/4015>, right?
> I should probably reread the fix again.
> Knowing that the data is the training data is indeed useful information.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4046#issuecomment-69107510>
> .
>

MechCoder commented:
Sorry for taking some time to respond.

> Say there are duplicates in the data, and I call the `X=None` variant of `radius_neighbors_graph`. I think I would like those duplicate points to be explicit zeros in the sparse matrix...

I am not sure that's the right way, because
a] It is inconsistent with the current behavior of `kneighbors_graph` (in this PR) which marks duplicates as neighbors, irrespective of whether X is None or not.
b] Technically, the duplicates should also be marked as neighbors, since they are within a radius distance. I think any such computation wrt clustering, could be done better outside since you have an idea about how many duplicates are there, instead of passing it to radius or kneighbors and doing the work again inside the function. 

> you'll get an error in the reshape!

Ouch. Yes this should be fixed. Your fix seems to be the best that I can think of too.

> This is bad because DRY.

We could always refactor while working towards a consistent API.

@amueller The question is how to determine if the query data is the same as the indexed data or not
We want to determine this because
1. if the query is the same as the indexed data, then it helps to skip the first NN as it most probably is the sample itself, (a corner-case is one given by @jnothman, when the number of duplicates is at least large as `n_samples/2`)
2. Otherwise it is better to not do anything special, and return the `k_neighbors` as they are.

Allowing the user to set `X=None` helps him (or her) realize easily, that the query is the same as the indexed data. Of course, we could also re introduce, `skip_first` which would do the same thing, but it is not that clear to anyone as to when this should be set True or False. (except maybe @jnothman and me, since we've discussed this over and over)

MechCoder commented:
Oh and I haven't fixed up `radius_neighbors` yet. I would like to as soon as this gets merged.

jnothman commented:
> I am not sure that's the right way

I think your points are in agreement with what I'm saying. I'm saying in distance mode we need a test to ensure that these are marked, and explicitly set in the sparse matrix even though their distance is 0 (which might otherwise be eliminated from a sparse matrix representation).

> number of duplicates is at least large as n_samples/2

I think the corner case was only when the number of duplicates is > n_neighbors

> we could also re introduce, skip_first

And I think we've made patently clear that "first" isn't our concern.

MechCoder commented:
Oh I'm sorry I misread, what you meant.

> I think the corner case was only when the number of duplicates is > n_neighbors.
Oops, I don't know where the `/2` came from.

> And I think we've made patently clear that "first" isn't our concern.
Yes, I've contradicted it in the next line :)

MechCoder commented:
> we need a test to ensure that these are marked, and explicitly set in the sparse matrix even though their distance is 0
 
Should we advice users to run `eliminate_zeros`if they want to avoid this? There might be cases where this might not be that useful as well, though (if someone decided to divide to blindly divide by `data` somewhere?)

jnothman commented:
Not sure `eliminate_zeros` helps anything. Please explain what use-case you mean?

MechCoder commented:
I'm not sure where this may be useful, but lets say somewhere one would want to scale down by `nn.data`

     In [19]: X = np.random.rand(5, 5)
    In [20]: X[1] = X[0]
    In [21]: nn = NearestNeighbors().fit(X).kneighbors_graph(None, 2, mode='distance')
    In [22]: nn.data
    Out[22]: 
    array([ 0.        ,  0.85527722,  0.        ,  0.85527722,  0.59708809,
            0.82572895,  0.65834477,  0.82572895,  0.59708809,  0.65834477])


Am I being too paranoid?

MechCoder committed:
FIX: Handle and test case where n_duplicates > n_neighbors

MechCoder commented:
@jnothman I have fixed the corner case (and tested for explicit zeros) in the last commit. @amueller , @jakevdp I would appreciate your reviews also. :)

jnothman commented:
If someone wants to scale down by the data, and has queries identical to
indexed points, that is something they'll have to deal with in their design.

On 9 January 2015 at 19:46, Coveralls <notifications@github.com> wrote:

>  [image: Coverage Status] <https://coveralls.io/builds/1715917>
>
> Coverage increased (+0.01%) when pulling *c9c7c55
> <https://github.com/scikit-learn/scikit-learn/commit/c9c7c550b7595bc6d16b1b1efefc65c833873f72>
> on MechCoder:allow_X_none* into *e98c513
> <https://github.com/scikit-learn/scikit-learn/commit/e98c513b31a9457734a0cc4972b4b8d8631410d5>
> on scikit-learn:master*.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4046#issuecomment-69307691>
> .
>

MechCoder commented:
Ok, So does everyone agree this PR does a fair job in handling duplicates ?

jnothman commented:
> Ok, So does everyone agree this PR does a fair job in handling duplicates ?

You know as well as we do that so far no one has looked at it! :)

jnothman commented:
But I aim to give you benchmarks of the improvement without parallelism on a real dataset I'm using.

MechCoder commented:
> You know as well as we do that so far no one has looked at it! :)

Just one more trick in my book to attract attention to review PR's ;)

> But I aim to give you benchmarks of the improvement without parallelism

That's great news!. But I seriously doubt if this was typed in the correct place.

jnothman commented:
Indeed.

jnothman commented:
I realise this is a fix, but I don't think we should allow `X=None` for `kneighbors` until the same is available for `radius_neighbors`. It is a fix to the API, so why leave the API inconsistent?

However, as a side note, I suspect the necessary manipulation in radius_neighbors can be more efficiently performed if `BinaryTree` directly returned the graph and `radius_neighbors` split the indices, rather than the other way (concatenation).

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -300,11 +300,26 @@ class from an array representing our data set and ask who's
         if n_neighbors is None:
             n_neighbors = self.n_neighbors
 
+        query_is_train = True
```
I'd rather this in the else case, even if flakes doesn't like it.

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -300,11 +300,26 @@ class from an array representing our data set and ask who's
+            n_neighbors += 1
+
+        n_samples, _ = X.shape
+        j = np.arange(n_samples)[:, None]
```
better name?

jnothman replied:
I don't get what j is for anyway. Can't you replace `j` below by `:`?

jnothman replied:
Not in the new logic, that is.

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -334,11 +350,39 @@ class from an array representing our data set and ask who's
+            # of neighbors, the first NN will not be the sample, but a duplicate.
+            # In that case mask the first duplicate.
+            dup_gr_nbrs = np.all(sample_mask, axis=1)
+            if np.any(dup_gr_nbrs):
```
Does this condition help anything? I think it's a waste of time and code.

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -334,11 +350,39 @@ class from an array representing our data set and ask who's
             raise ValueError("internal: _fit_method not recognized")
 
-    def kneighbors_graph(self, X, n_neighbors=None,
+        if query_is_train:
```
Put the shorter case first:
```
if not query_is_train:
    return result
else:
    ...
```
(I'm also okay with having no else but I've seen that disperefered in the project)

jnothman commented:
Docstrings?

Note to self: not looked at tests yet.

jnothman commented:
> Just one more trick in my book to attract attention to review PR's ;)

But really, you had to nag me before you added docstrings??

MechCoder commented:
@jnothman Thanks. Should I fix `radius_neighbors_graph` up before anyone else has a look?

MechCoder committed:
Add docstring and test for kneighbors

amueller commented:
I'll try to review before the end of the week.

jnothman commented:
I still think it's awkward leaving them with inconsistent API. Maybe I'll find the moment to patch up radius_neighbors equivalently.

MechCoder commented:
It's alright, I will push in a commit in some time.

MechCoder commented:
A problem with refactoring is that the `radius_neighbors` sometimes returns an object array, so the same approach of masking and then reshaping will not work.We would probably need to iterate across `neigh_ind` and mask each sample individually.

MechCoder commented:
So the best way to prevent thie reiteration, would be to handle this in `neigh_ind` computation.

jnothman commented:
> radius_neighbors sometimes returns an object array

It should always return an object array.

I think don't worry about the iteration. It's fine for now. Already any
approach that uses it needs to iterate through a collection of objects. The
only alternative I can consider would be to construct the CSR graph
(perhaps in the BinaryTree) and operate on that.

On 15 January 2015 at 16:17, Manoj Kumar <notifications@github.com> wrote:

>  So the best way to prevent thie reiteration, would be to handle this in
> neigh_ind computation.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4046#issuecomment-70040978>
> .
>

MechCoder commented:
I should have phrased it the other way (https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/neighbors/base.py#L491)

It returns an int array when possible

jnothman commented:
Oh, perhaps I only proposed changing that. It makes for very surprising
behaviour, unless the number of queries is 1, which I guess may be best
handled in such a way that np.squeeze works.

On 15 January 2015 at 16:26, Manoj Kumar <notifications@github.com> wrote:

>  I should have phrased it the other way (
> https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/neighbors/base.py#L491
> )
>
> It returns an int array when possible
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4046#issuecomment-70041503>
> .
>

MechCoder commented:
That seemed to sound good, but there is what I think is a problem with how NumPy handles object arrays.

    In [28]: neigh = [[1], [2],[3], [4]]
    In [29]: i = np.array(neigh, dtype='object')
    # Should this have returned a list?
    In [30]: i[0]
    Out[30]: array([1], dtype=object)
    In [31]: i[0][0]
    Out[31]: 1

This would cause an error here, https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/neighbors/base.py#L499 since NumPy indexing does not support object arrays. One way I can think of is doing a dtype check and handle the cases separately.

MechCoder commented:
Just fyi: I reported it in NumPy here, https://github.com/numpy/numpy/issues/5456

amueller commented on sklearn/neighbors/graph.py:
```diff
@@ -77,7 +79,13 @@ def kneighbors_graph(X, n_neighbors, mode='connectivity', metric='minkowski',
         _check_params(X, metric, p, metric_params)
-    return X.kneighbors_graph(X._fit_X, n_neighbors, mode=mode)
+    if mode == "connectivity":
+        warnings.warn("The behavior of marking each sample itself as the "
```
Maybe say "is deprecated since 0.16 and will be changed in 0.18" or maybe also say what it will change to?

jnothman replied:
I don't know if it makes sense to say that the behaviour is deprecated, given that it remains unchanged. I do think you need to state that for future-proof code, the `NearestNeighbors.kneighbors_graph` is the recommended alternative.

jnothman replied:
Also, these warnings should ideally be asserted for, or at least ignored, where they are triggered in tests.

MechCoder replied:
Are you sure? As far as I understand, we deprecate two versions earlier, to warn the user that the source code he is working with is not the same and is going to change two versions later, so her better change his applications that use this part of the source code (or else face the consequences)

jnothman replied:
I guess what I mean is that the API usually gets deprecated: we warn that something that used to be possible will no longer be available once complete. Here, it is not that something will become available, it's that a behaviour will change. I don't think of that as deprecation.

MechCoder replied:
So what do we do? Remove the warning and document this as a bug-fix?

jnothman replied:
I have no problem with the word warning, I only find the word "deprecated" confusing, and the lack of advice on what to do instead.

amueller replied:
Ok, let's not use the word deprecated. But I am still not sure on what a good plan of action is. We want to keep this method and have it be useful in the future, right? And we don't want to change current behavior. But it should also be useful now. If it is not useful now we can basically deprecate the whole function and remove it in the future, because everybody needs to switch away from it because it is broken.
So what I would like to have is a future-proof way for people to use the fixed version.

amueller replied:
The only way I could think of is adding an additional parameter, but then we would have to deprecate the parameter again. That seems overly complicated, and means the user needs to change their code twice, once adding the ``i_want_the_fix=True`` and then removing it again.

MechCoder replied:
Hmm.. But what is wrong in warning the user that the behavior is going to change in 0.18 and if he wants to make sure the code is future proof, then it is wise of him to replace it with `NearestNeighbors().fit(X)`?

MechCoder replied:
Apart from the huge maintenance cost, of course.

amueller replied:
Let's say someone now writes new code, sees the method, and wants to use it.
Then they get a warning, telling them
 "this method is broken" and they will have to use a different method. Then why keep this method at all if they can't use it now?

amueller replied:
With this fix we are basically saying "this method is unusable for two releases". I don't like that at all.

jnothman replied:
Except perhaps in 'distance' mode. I think I agree. Let's deprecate the whole method and remove it, hopefully with @jakevdp's blessing.

amueller replied:
Just to be clear: I was not entirely suggesting to remove the method.
I would prefer to keep the method but make it useable now. I just don't see a good way to achieve that.

[I was talking about the connectivity mode]

MechCoder replied:
If we all agree that `kneighbors_graph(X, params)` is just the lazy guy's way of writing `NearestNeighbors(params).fit(X).kneighbors_graph()` and if we want the behavior to be non-buggy, I think that is the best way to go.

amueller commented:
I'm only now really trying to catch up on this.
So just to summarize and clarify the motivation and desired behavior:
This issue fixes an inconsistency between
``NearestNeighbors.kneighbors_graph(mode="distance")`` and ``NearestNeighbors.kneighbors_graph(mode="connectivity")``, which also resulted in an inconcistency
 between ``kneighbors_graph(mode="distance")`` and ``kneighbors_graph(mode="connectivity")`` (#4015)
Also ``NearestNeighbors.kneighbors_graph(mode="distance")`` was buggy / really odd, when not used with the training data.

``kneighbors_graph`` always works on one dataset, but uses NearestNeighbors in it's implementation, which does not keep track of the information whether the query is the training data. So to implement ``kneighbors_graph`` in a sensible way, we need to introduce an option to ``NearestNeighbors.kneighbors_graph`` to communicate that X is the training data, which is ``X=None``. This makes ``kneighbors_graph`` work sensible for ``mode="connectivity"``.

With this new option, it is also clear that when ``X is not None`` in ``NearestNeighbors.kneighbors_graph``, then it is not (meant as) the training data, so identical points should be connected. That changes the behavior of ``mode="distance"`` slightly.

To me it looks like in the PR
1) You added ``X=None`` behavior to ``NearestNeighbors.kneighbors`` but didn't change any existing behavior there. 
2) You fixed the weird ``NearestNeighbors.kneighbors_graph(mode="distance")`` behavior which was quite odd / buggy.
3) You added a note to ``kneighbors_graph(mode="connectivity")`` that it will be more sensible in the future.

Understanding this now, this looks awesome :) [and sorry for the long comment].
Just to make sure, the behavior of this:
``NearestNeighbors().fit(X).kneighbors_graph(X, mode="distance")``
is now slightly different in having explicit zeros instead of no entries on the diagonal, right? And we hope that doesn't break peoples code...

for the fix to  ``kneighbors_graph(mode="connectivity")``, I think it would be nice to either allow people to already use the fix, or maybe more simply to tell them how to get the correct version.
The current behavior does not allow them to use it without getting a warning, which is odd.

amueller commented on sklearn/neighbors/base.py:
```diff
@@ -300,11 +303,27 @@ class from an array representing our data set and ask who's
+        else:
+            query_is_train = True
+            X = self._fit_X
+            # Include an extra neighbor in order to avoid marking the sample
```
that is not to avoid this behavior, right? It is to remove this one later on.

jnothman replied:
It's a confusing comment, whatever intended. "Include an extra neighbor to account for the sample itself being returned, which is removed later" or something might be clearer.

amueller commented on sklearn/neighbors/base.py:
```diff
@@ -334,18 +355,48 @@ class from an array representing our data set and ask who's
+            # of neighbors, the first NN will not be the sample, but a duplicate.
+            # In that case mask the first duplicate.
+            dup_gr_nbrs = np.all(sample_mask, axis=1)
+            sample_mask[:, 0][dup_gr_nbrs] = False
```
I would probably have masked the last one but it is an arbitrary choice anyhow ^^

amueller commented on sklearn/neighbors/base.py:
```diff
@@ -379,33 +430,39 @@ def kneighbors_graph(self, X, n_neighbors=None,
-                                        return_distance=True)
-            A_data, A_ind = data[:, 1:], ind[:, 1:]
+            A_data, A_ind = self.kneighbors(X, n_neighbors, return_distance=True)
+            A_data = np.ravel(A_data)
```
why did you move that one to a separate line?

MechCoder replied:
I have the A_data in `mode=connectivity` case, to prevent creating a 2-D array and then ravelling at the end. Hence.

amueller replied:
oh, sorry, I see.

amueller commented on sklearn/neighbors/tests/test_neighbors.py:
```diff
@@ -895,14 +895,90 @@ def test_non_euclidean_kneighbors():
     X_nbrs = neighbors.NearestNeighbors(3, metric='manhattan')
     X_nbrs.fit(X)
-    assert_raises(ValueError, neighbors.kneighbors_graph, X_nbrs, 3,
+    assert_raises(ValueError, neighbors.kneighbors_graph, None, 3,
```
That is actually a funny behavior, isn't it?

amueller commented:
The code looks good, but I have one concern: it feels like this is a somewhat hackish fix, and querying the training data should be supported on a lower level.
In particular, if I build a ``NearestNeighbors`` object with ``n_neighbors=3`` I can not query the three closest neighbors for a training point. Does the underlying implementation not allow for that? That seems very odd. Maybe @jakevdp can say.

amueller commented on sklearn/neighbors/tests/test_neighbors.py:
```diff
@@ -895,14 +895,90 @@ def test_non_euclidean_kneighbors():
+        assert_array_equal(ind, [[1], [0]])
+
+        # Do not do anything special to duplicates.
+        kng = nn.kneighbors_graph([[0], [1]], mode='distance')
```
maybe loop over "mode", too, to reduce code duplication?

amueller commented on sklearn/neighbors/tests/test_neighbors.py:
```diff
@@ -895,14 +895,90 @@ def test_non_euclidean_kneighbors():
+        kng = nn.kneighbors_graph([[0], [1]], mode='distance')
+        assert_array_equal(
+            kng.A,
+            np.array([[0., 0.], [0., 0.]]))
```
Not to self and other reviewers: this is a non-regression test that wouldn't pass on master...

amueller commented on sklearn/neighbors/tests/test_neighbors.py:
```diff
@@ -895,14 +895,90 @@ def test_non_euclidean_kneighbors():
+        assert_array_equal(kng.data, [0., 0.])
+        assert_array_equal(kng.indices, [0, 1])
+
+        assert_array_equal(
```
thisalso doesn't pass on master...

amueller commented:
Apart from the fact that I feel the impact on the ``n_neighbors`` parameter is odd, and that I feel this might not be the right implementation, the interface, code and tests look good. Great job on iterating on this :)

amueller commented:
Never mind my concern about ``n_neighbors``. I thought this was actually used in the tree construction for some reason. This all looks good.
+1 for the current changes. I don't have a strong opinion on whether the radius neighbors should also be fixed in this PR.

MechCoder committed:
ENH/FIX: Patch radius_neighbors to follow kneighbors convention

MechCoder commented:
@jnothman I've patched up the `radius_neighbors` too, so hopefully it will be less awkward :P . @amueller thanks for your comments. I will have a look now.

jnothman commented:
> In particular, if I build a NearestNeighbors object with n_neighbors=3 I can not query the three closest neighbors for a training point. 

I'm not exactly sure what you meant of this struck-out comment, but I've realised it's a bit annoying that you can't query with only a portion of the training data and get the same behaviour as X=None... but in any case to do so you would need to provide the method with which indices to avoid for which query point, so the user might as well just handle that case anyway.

jnothman commented:
> @jnothman I've patched up the radius_neighbors too, so hopefully it will be less awkward :P .

Thanks, I'll try to give it a full look-over soon.

jnothman commented:
Note: Travis is unhappy

MechCoder commented:
Yes, Its a doctest that seems to give different results in different versions of NumPy (related with the object handling) (passes on other two builds)

jnothman commented:
Yes, arrays of arrays show differently in different versions of numpy. Perhaps just extract an element in the doctest, or else use ellipsis.

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -300,11 +303,27 @@ class from an array representing our data set and ask who's
+        train_size = self._fit_X.shape[0]
+        if n_neighbors > train_size:
+            raise ValueError(
+                "Expected n_neighbors <= %d. Got %d" % (train_size, n_neighbors)
```
Line too long

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -300,11 +303,27 @@ class from an array representing our data set and ask who's
+                "Expected n_neighbors <= %d. Got %d" % (train_size, n_neighbors)
+            )
+        n_samples, _ = X.shape
+        sample_range = np.arange(n_samples)[:, None]
```
You moved this away from the locality in which it was used. You can move it back now...

MechCoder replied:
I do not understand, this is used in all places except the `kd_tree` and `ball_tree` methods.

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -334,18 +355,48 @@ class from an array representing our data set and ask who's
+            sample_mask = neigh_ind != sample_range
+
+            # Corner case: When the number of duplicates are more than the number
+            # of neighbors, the first NN will not be the sample, but a duplicate.
```
More line too long. Please PEP8

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -334,18 +355,48 @@ class from an array representing our data set and ask who's
+                neigh_ind[sample_mask], (n_samples, n_neighbors - 1))
+
+            if return_distance:
+                dist = np.reshape(dist[sample_mask], (n_samples, n_neighbors - 1))
```
Line too long

jnothman commented on sklearn/neighbors/tests/test_neighbors.py:
```diff
@@ -895,14 +895,136 @@ def test_non_euclidean_kneighbors():
                   radius, metric='euclidean')
 
 
+def test_kneighbors_graph_nearest_neighbors():
```
I think you just mean `test_kneighbors_graph_method`.

But I suspect you should really have a test of the X=None case (i.e. `def_test_neighbors_no_query` or some such), and ensure that the tests of the X != case elsewhere are sufficient; then the tests of the graph methods should really just be checking equivalence between the graph and non-graph methods for various values of `X_{train,test}`.

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -334,18 +355,48 @@ class from an array representing our data set and ask who's
+            else:
+                neigh_ind = result
+
+            # If the query data is the same as the indexed data, we would like
```
Put this right after "else:", IMO

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -514,13 +578,33 @@ class from an array representing our data set and ask who's
 
+        if not query_is_train:
+            return results
+        else:
```
Insert comment to say you're removing the query index.

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -514,13 +578,33 @@ class from an array representing our data set and ask who's
+
+            for ind, ind_neighbor in enumerate(neigh_ind):
+                mask = ind_neighbor != ind
+                # Corner case if the number of duplicates is more than the number
```
This case is inapplicable. There's no such thing as "number of neighbors", and the query index should always be there (at least for eps > 0, and = 0 given an appropriate resolution of #4072 ). (I'd be very surprised if the line in the `if` below has test coverage!!!)

MechCoder replied:
Ah yes, you are right!

amueller replied:
I was very confused by this until I saw that this is in radius neighbors lol.

MechCoder replied:
Oh come on, it was Saturday morning when I wrote this :(

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -563,14 +647,17 @@ def radius_neighbors_graph(self, X, radius=None, mode='connectivity'):
         kneighbors_graph
         """
-        X = check_array(X, accept_sparse=['csr', 'csc', 'coo'])
+        # radius_neighbors does the None handling.
```
I think that's not really deserving of comment.

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -563,14 +647,17 @@ def radius_neighbors_graph(self, X, radius=None, mode='connectivity'):
+        # radius_neighbors does the None handling.
+        if X is not None:
+            X = check_array(X, accept_sparse='csr')
+            n_samples1 = X.shape[0]
```
Just get this value from `A_ind` below (in either case).

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -563,14 +647,17 @@ def radius_neighbors_graph(self, X, radius=None, mode='connectivity'):
-        X = check_array(X, accept_sparse=['csr', 'csc', 'coo'])
+        # radius_neighbors does the None handling.
+        if X is not None:
+            X = check_array(X, accept_sparse='csr')
```
Do we have any idea why CSC/COO were previously supported? I think since calculations are row-wise, constraining to CSR makes sense, but changing it should not be arbitray.

MechCoder committed:
TST: Refactor tests for k and radius neighbors

MechCoder commented:
@jnothman I have addressed all your comments. @amueller Did you mean something like this? 

"Use NearestNeighbors.fit(X).kneighbors_graph(X) if you want to retain the same behavior"?

jnothman commented on sklearn/neighbors/tests/test_neighbors.py:
```diff
@@ -895,14 +895,134 @@ def test_non_euclidean_kneighbors():
+        X = [[0], [1]]
+        nn.fit(X)
+        test_data = [[2], [1]]
+        assert_array_equal(
```
I still think testing `*_graph` methods here is getting in the way, when you can just test that for various `X`, the `_graph` variants have a fixed relationship to the base forms.

MechCoder replied:
But what if someone adds something to these methods, so that they don't have fixed relationships anymore tomorrow? I'll put these in separate functions, but I think they should be there.

jnothman replied:
Well you'd still test that fixed relationship is true of a number of input datasets, including the ones you're testing with here, so you'd be asserting no less, just it would be more straightforward to read.

MechCoder replied:
But we should still test the explicit marking of zeros in the graph variants.

jnothman commented on sklearn/neighbors/tests/test_neighbors.py:
```diff
@@ -895,14 +895,134 @@ def test_non_euclidean_kneighbors():
+            nn.kneighbors_graph().A,
+            np.array([[0., 1., 1.], [1., 0., 1.], [1., 1., 0]]))
+
+        # Mask the first duplicates when n_duplicates > n_neighbors.
```
Should these thus be in "test_k_and_radius_neighbors_duplicates"?

MechCoder commented:
@amueller I see that you tagged this as 0.16 milestone. When is the scheduled date?

amueller commented:
!(soon)[http://www.hitmanforum.com/uploads/default/918/e94152c2d4e3d535.png]
A week or two, I think.

amueller commented:
On 01/18/2015 01:22 AM, Manoj Kumar wrote:
>
> @jnothman <https://github.com/jnothman> I have addressed all your 
> comments. @amueller <https://github.com/amueller> Did you mean 
> something like this?
>
> "Use NearestNeighbors.fit(X).kneighbors_graph(X) if you want to retain 
> the same behavior"?
>
>
Yes.

MechCoder commented:
@jnothman @amueller Should I push a commit that does the deprecation, or should I wait. It seems like no one else is following this Pull Request. And it should take some time for anyone to understand what is going on.

amueller commented:
I think we should wait a bit and maybe hope that @jakevdp or @ogrisel find the time.
Deprecating public API is something I don't like to do lightly.

amueller commented:
@MechCoder you also fixed this, right: #1400?

MechCoder commented:
Yes, but not intentionally :D , now radius_neighbors returns an object array irrespective of whatever method. Should I post to the mailing list to attract some attention.

Basically I was working on some other issue fix and I ran into this, so I would like to see this in.

jnothman commented:
The same behaviour needs to be true of BallTree and KDTree; and of
LSHForest.

On 23 January 2015 at 16:06, Manoj Kumar <notifications@github.com> wrote:

>  Yes, but not intentionally :D , now radius_neighbors returns an object
> array irrespective of whatever method. Should I post to the mailing list to
> attract some attention.
>
> Basically I was working on some other issue fix and I ran into this, so I
> would like to see this in.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4046#issuecomment-71148359>
> .
>

MechCoder commented:
I see, is it okay to do this in a separate branch. (I find smaller PR's easier to interview than one huge PR)

amueller commented:
yes, I also prefer smaller PRs. We should get this merged but I'm still not 100% sure on what do to with the functions.

amueller commented:
Would you mind explaining the situation on the ML and asking whether we should keep the functions or deprecate them?

amueller commented:
After thinking about it for some time, I think we should introduce an ``include_self`` parameter to the ``kneighbors_graph`` function, make it default to ``None`` and if it is ``None`` warn that the default behavior will change from ``True`` to ``False`` in two versions.

jnothman commented:
Thanks for thinking this through. I'm happy with that.

On 27 January 2015 at 11:47, Andreas Mueller <notifications@github.com>
wrote:

>  After thinking about it for some time, I think we should introduce an
> include_self parameter to the kneighbors_graph function, make it default
> to None and if it is None warn that the default behavior will change from
> True to False in two versions.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4046#issuecomment-71569843>
> .
>

MechCoder commented:
Which should be only for `mode=connectivity` right?

jnothman commented:
Why so?

On 27 January 2015 at 17:59, Manoj Kumar <notifications@github.com> wrote:

>  Which should be only for mode=connectivity right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4046#issuecomment-71598402>
> .
>

MechCoder commented:
I had meant that presently that `mode=distance` does not `include_self` by default, i.e the current behavior itself is correct, so we might not need to raise a warning for that. wdyt?

amueller commented:
If we include the parameter anyhow, I think we can make a proper backward compatible change. The difference for ``mode=distance`` is not that much anyhow.

MechCoder commented:
I am for including the parameter. My question was just about the warning.

jnothman commented:
> I had meant that presently that mode=distance does not include_self by
default, i.e the current behavior itself is correct, so we might not need
to raise a warning for that. wdyt?

No, the current behaviour is broken. It arbitrarily removes the first
matched. But also, that's a matter of where the user is aware of explicit
zeros. I think keep it consistent between the two modes; the user gets a
warning that things have changed as well, and if especially concerned, you
could add to that warning something about the changed behaviour.

On 28 January 2015 at 08:10, Manoj Kumar <notifications@github.com> wrote:

>  I am for including the parameter. My question was just about the warning.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4046#issuecomment-71728586>
> .
>

MechCoder committed:
ENH: Add include_self param to graph variants for backward compat

MechCoder commented:
@jnothman @amueller Fixed. Anything more other than a `whatsnew` entry?

amueller commented on sklearn/neighbors/graph.py:
```diff
@@ -77,11 +103,13 @@ def kneighbors_graph(X, n_neighbors, mode='connectivity', metric='minkowski',
 
-def radius_neighbors_graph(X, radius, mode='connectivity', metric='minkowski',
-                           p=2, metric_params=None):
+def radius_neighbors_graph(X, radius, include_self=None, mode='connectivity',
```
I would always add new parameters to the end, in case anyone was relying on parameter order.

amueller commented:
Apart from my comment :+1: for merge. Thanks for your perseverance!

MechCoder committed:
DOC: Add what's new entry

MechCoder commented:
Thanks, can you have a quick look at the what's new doc? Anything more that I need to add?

amueller commented:
Looks good!

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -247,15 +247,18 @@ def _fit(self, X):
         Parameters
         ----------
-        X : array-like, last dimension same as that of fit data
+        X : array-like, last dimension same as that of fit data or None.
```
"optional"

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -247,15 +247,18 @@ def _fit(self, X):
         ----------
-        X : array-like, last dimension same as that of fit data
+        X : array-like, last dimension same as that of fit data or None.
             The new point.
```
"The query point or points"

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -334,18 +356,50 @@ class from an array representing our data set and ask who's
         ----------
-        X : array-like, shape = [n_samples, n_features]
-            Sample data
+        X : array-like, last dimension same as that of fit data or None.
```
"optional" rather than "or None".

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -379,39 +433,45 @@ def kneighbors_graph(self, X, n_neighbors=None,
 
-        return csr_matrix((A_data.ravel(), A_ind.ravel(), A_indptr),
-                          shape=(n_samples1, n_samples2))
+        kneighbors_graph = csr_matrix((
```
This doesn't look like correct indentation. The first entry in the first argument tuple shouldn't be lining up with the kwarg name. What's wrong with:
```
        kneighbors_graph = csr_matrix((A_data, A_ind.ravel(), A_indptr),
                                      shape=(n_samples1, n_samples2))
```

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -419,7 +479,10 @@ def radius_neighbors(self, X, radius=None, return_distance=True):
         Parameters
         ----------
         X : array-like, last dimension same as that of fit data
```
"optional"

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -419,7 +479,10 @@ def radius_neighbors(self, X, radius=None, return_distance=True):
         ----------
         X : array-like, last dimension same as that of fit data
-            The new point or points
+            The new point or points.
```
"The query point or points."

jnothman commented on sklearn/neighbors/base.py:
```diff
@@ -419,7 +479,10 @@ def radius_neighbors(self, X, radius=None, return_distance=True):
         X : array-like, last dimension same as that of fit data
-            The new point or points
+            The new point or points.
+            If not provided, X is assumed to be the same as the fit
```
Perhaps, "If not provided, neighbors of each indexed [or: training] point are returned." I'm not sure, then that it needs to explicitly state something like "In this case, the query point is not considered its own neighbor."

jnothman commented on sklearn/neighbors/graph.py:
```diff
@@ -20,8 +22,27 @@ def _check_params(X, metric, p, metric_params):
+        warnings.warn(
+            "include_self will be set to False, i.e the first NN of each "
+            "sample will not be the sample itself unless explicitly set "
+            "from version 0.18", DeprecationWarning)
```
perhaps "Please specify include_self=True if this behavior is not sought."?

jnothman commented on sklearn/neighbors/graph.py:
```diff
@@ -44,6 +65,11 @@ def kneighbors_graph(X, n_neighbors, mode='connectivity', metric='minkowski',
         The default distance is 'euclidean' ('minkowski' metric with the p
         param equal to 2.)
 
+    include_self: bool, default None
```
"default True"

MechCoder replied:
The default option is not True. In mode distance, the original behavior was include_self = False

MechCoder replied:
Or did you mean in the future?

jnothman replied:
Hmmm... I'm almost inclined to not do that, because the distance mode was include_self=broken, not False. But you're right. Still "None" tells the user nothing. Tell them instead, default is backwards-compatible, or default=True iff mode='connectivity'.

jnothman commented:
Apart from making the docs a bit more intelligible, I'm +1

jnothman commented on sklearn/neighbors/graph.py:
```diff
@@ -20,8 +22,27 @@ def _check_params(X, metric, p, metric_params):
+    """Return the query based on include_self param"""
+    # Done to preserve backward compatibility.
+    if include_self is None:
+        warnings.warn(
```
Warning's not necessary if mode='distance'

MechCoder replied:
You were the one who convinced me the other way.

jnothman replied:
I'm glad I'm persuasive, but I'm sure I never intended that.

MechCoder committed:
DOC: Make docs clearer

MechCoder commented:
@jnothman I've fixed up your comments. Can you have a last look?

jnothman commented:
LGTM, then.

MechCoder commented:
Alright, merging then. Thanks for your reviews and patience.

amueller commented:
Wohoo, thanks :)
